vault/identity_store_oidc_provider.go

```diff
@@ -1060,3 +1179,461 @@ func (i *IdentityStore) pathOIDCProviderExistenceCheck(ctx context.Context, req
+		entityID: entity.GetID(),
+		nonce:    nonce,
+		scopes:   scopes,
+	}
+
+	// Validate the optional max_age parameter to check if an active re-authentication
+	// of the user should occur. Re-authentication will be requested if max_age=0 or the
+	// last time the token actively authenticated exceeds the given max_age requirement.
+	// Returning ErrAuthMaxAgeReAuthenticate will enforce the user to re-authenticate via
+	// the user agent.
+	if maxAgeRaw, ok := d.GetOk("max_age"); ok {
+		maxAge := maxAgeRaw.(int)
+		if maxAge < 0 {
+			return logical.ErrorResponse("%s: max_age must be greater than zero",
+				ErrAuthInvalidRequest), nil
+		}
+		if maxAge == 0 {
+			// TODO: solve for the potential UI loop here or make max_age=0 invalid
+			return logical.ErrorResponse("%s: active re-authentication is required by max_age",
+				ErrAuthMaxAgeReAuthenticate), nil
+		}
+
+		// Look up the token associated with the request
+		te, err := i.tokenStorer.LookupToken(ctx, req.ClientToken)
+		if err != nil {
+			return nil, fmt.Errorf("%s: internal server error", ErrAuthServerError)
+		}
+		if te == nil {
+			return logical.ErrorResponse("%s: token associated with request not found",
+				ErrAuthAccessDenied), nil
+		}
+
+		// Check if the token creation time violates the max age requirement
+		now := time.Now().UTC()
+		lastAuthTime := time.Unix(te.CreationTime, 0).UTC()
+		secondsSince := int(now.Sub(lastAuthTime).Seconds())
+		if secondsSince > maxAge {
+			return logical.ErrorResponse("%s: active re-authentication is required by max_age",
+				ErrAuthMaxAgeReAuthenticate), nil
+		}
+
+		// Set the auth time to use for the auth_time claim in the token exchange
+		authCodeEntry.authTime = lastAuthTime
+	}
+
+	// Generate the authorization code
+	code, err := base62.Random(32)
+	if err != nil {
+		return nil, fmt.Errorf("%s: internal server error", ErrAuthServerError)
+	}
+
+	// Cache the authorization code for a subsequent token exchange
+	if err := i.oidcAuthCodeCache.SetDefault(ns, code, authCodeEntry); err != nil {
+		return nil, fmt.Errorf("%s: internal server error", ErrAuthServerError)
+	}
+
+	return &logical.Response{
+		Data: map[string]interface{}{
+			"code": code,
+		},
+	}, nil
+}
+
+// entityHasAssignment returns true if the entity is a member of any of the
+// assignments' groups or entities. Otherwise, returns false or an error.
+func (i *IdentityStore) entityHasAssignment(ctx context.Context, s logical.Storage, entity *identity.Entity, assignments []string) (bool, error) {
+	for _, a := range assignments {
+		assignment, err := i.getOIDCAssignment(ctx, s, a)
+		if err != nil {
+			return false, fmt.Errorf("%s: internal server error", ErrAuthServerError)
+		}
+		if assignment == nil {
+			return false, fmt.Errorf("%s: client assignment %q not found", ErrAuthServerError, a)
+		}
+
+		// Get the group names that the entity is a member of
+		entityGroups, err := i.MemDBGroupsByMemberEntityID(entity.GetID(), true, false)
```

Comment:
Same question here as above. Is it possible for the Groups to exist in Storage but not MemDB? If yes, would that be an issue here?

Reply:
See https://github.com/hashicorp/vault/pull/12538#discussion_r709606050. Checked that this is also true for `MemDBGroupsByMemberEntityID`.